lua/starfall/libs_sv/entities.lua

```diff
@@ -30,6 +30,156 @@ registerprivilege("entities.canTool", "CanTool", "Whether or not the user can us
 registerprivilege("entities.use", "Use", "Whether or not the user can use the entity", { entities = {} })
 registerprivilege("entities.getTable", "GetTable", "Allows the user to get an entity's table", { entities = {}, usergroups = { default = 1 } })
 
+
+local collisionListenerLimit = SF.LimitObject("collisionlistener", "collisionlistner", 200, "The number of concurrent starfall collision listeners")
+local base_physicscollide
+SF.GlobalCollisionListener = {
+	__index = {
+		create = function(self, ent)
+			local listenertable = {}
+		
+			local queue = {}
+			local function collisionQueueProcess()
+				if IsValid(ent) then
+					for _, listener in ipairs(listenertable) do
+						local instance = listener.instance
+						for i=1, #queue do
+							listener:run(SF.StructWrapper(instance, queue[i], "CollisionData"))
+						end
+					end
+				end
+				for i=1, #queue do
+					queue[i] = nil
+				end
+			end
+		
+			local function collisionQueueCallback(ent, data)
+				local i = #queue+1
+				queue[i] = data
+				if i==1 then timer.Simple(0, collisionQueueProcess) end
+			end
+		
+			if ent:IsScripted() then
+				local oldPhysicsCollide = ent.PhysicsCollide or base_physicscollide
+				ent.SF_OldPhysicsCollide = oldPhysicsCollide
+		
+				function ent:PhysicsCollide(data, phys)
+					oldPhysicsCollide(self, data, phys)
+					collisionQueueCallback(self, data)
+				end
+			else
+				ent.SF_CollisionCallback = ent:AddCallback("PhysicsCollide", collisionQueueCallback)
+			end
+			SF.CallOnRemove(ent, "RemoveCollisionListeners", function(e) self:destroy(e) end)
+
+			self.listeners[ent] = listenertable
+			return listenertable
+		end,
+		destroy = function(self, ent)
+			local entlisteners = self.listeners[ent]
+			if entlisteners==nil then return end
+			self.listeners[ent] = nil
+			for _, listener in ipairs(entlisteners) do
+				listener.manager:free(ent, listener)
+			end
+			if IsValid(ent) then
+				local oldPhysicsCollide = ent.SF_OldPhysicsCollide
+				if oldPhysicsCollide then
+					ent.PhysicsCollide = oldPhysicsCollide
+				else
+					ent:RemoveCallback("PhysicsCollide", ent.SF_CollisionCallback)
+				end
+		
+				SF.RemoveCallOnRemove(ent, "RemoveCollisionListeners")
+			end
+		end,
+		add = function(self, ent, listener)
+			local entlisteners = self.listeners[ent]
+			if entlisteners == nil then
+				entlisteners = self:create(ent)
+			elseif table.HasValue(entlisteners, listener) then
+				return
+			end
+			entlisteners[#entlisteners + 1] = listener
+		end,
+		remove = function(self, ent, listener)
+			local entlisteners = self.listeners[ent]
+			if entlisteners==nil then return end
+			local i = table.HasValue(entlisteners, listener)
```

Comment:
Need to see what the glua function for table.IndexOf is.

Reply:
table.KeyFromValue()